source/PlayState.hx

```diff
@@ -1,16 +1,187 @@
+			
+			var wall:Wall = new Wall(WORLD_LEFT-32,y);
+			var wall2:Wall = new Wall(WORLD_RIGHT,y);
+			walls.add(wall);
+			walls.add(wall2);
+		}
+		
+		add(walls);
+	}
+
+	private function initializeLava(){
+		lavaPools = new FlxTypedGroup<Lava>();
+
+		for (i in 0...LAVA_COUNT) {
+			var x:Float = LAVA_START_X + (i * Lava.WIDTH);
+			var y:Float = LAVA_START_Y;
+			var lava:Lava = new Lava(x, y);
+			lavaPools.add(lava);
+		}
+			add(lavaPools);
+	}
+
+	private function initializeFireballs() {
+		fireballs = new FlxTypedGroup<Fireball>();
+
+		for (i in 0...FIREBALL_COUNT) {
+			var x:Float = FlxG.random.int(FIREBALL_SPAWN_BORDER, 
+				FlxG.width - FIREBALL_SPAWN_BORDER);
+			var y:Float = FlxG.random.int(FIREBALL_SPAWN_BORDER, 
+				FlxG.height - FIREBALL_SPAWN_BORDER);
+			var fireball = new Fireball(x, y);
+			fireballs.add(fireball);
+		}
+		add(fireballs);
 	}
 
-	override public function update(elapsed:Float)
+	override public function update(elapsed:Float):Void
 	{
 		super.update(elapsed);
+
+
+		if(fireballs.length<FIREBALL_COUNT){
+			fireballs = new FlxTypedGroup<Fireball>();
+
+			var x:Float = FlxG.random.int(FIREBALL_SPAWN_BORDER, 
+				FlxG.width - FIREBALL_SPAWN_BORDER);
+			var y:Float = FlxG.random.int(FIREBALL_SPAWN_BORDER, 
+				FlxG.height - FIREBALL_SPAWN_BORDER);
+			var fireball = new Fireball(x, y);
+			fireballs.add(fireball);
+		}
+		
+
+		// Uncomment to collide hero against all wall objects.w
+		FlxG.collide(hero, walls);
+
+		FlxG.overlap(hero,lavaPools,resolveHeroLavaOverlap);
+
+		FlxG.overlap(hero,coin,resolveHeroCoinOverlap);
+
+		// Uncomment to trigger custom logic when a player overlaps with a fireball.
+		FlxG.overlap(hero, fireballs, resolveHeroFireballOverlap);
+
+		// Wrap various objects if gone offscreen.
+		// screenWrapObject(hero);
+		for (fireball in fireballs) {
+			screenWrapObject(fireball);
+		}
+		score.x = hero.x - FlxG.width + 100;
+		score.y = hero.y - FlxG.height + 100;
+
+	}
+
+	private function screenWrapObject(wrappingObject:FlxObject) {
+		if (wrappingObject.x > WORLD_RIGHT) {
+			wrappingObject.x = WORLD_LEFT - wrappingObject.width;
+		} else if (wrappingObject.x + wrappingObject.width < WORLD_LEFT) {
+			wrappingObject.x = WORLD_RIGHT;
+		}
+
+		if (wrappingObject.y > WORLD_BOTTOM) {
+			wrappingObject.y = WORLD_TOP - wrappingObject.height;
+		} else if (wrappingObject.y + wrappingObject.height < WORLD_TOP) {
+			wrappingObject.y = WORLD_BOTTOM;
+		}
+	}
+
+	/**
+		Function called when an overlap between hero and fireball is detected.
+	**/
+	private function resolveHeroFireballOverlap(hero:Hero, fireball:Fireball) {
+		fireball.kill();
+		hero.kill();
+		FlxG.switchState(new MenuState());
+	}
+
+	private function resolveHeroLavaOverlap(hero:Hero, lava:Lava){
+		hero.kill();
+		FlxG.switchState(new MenuState());
+	}
+
+	private function resolveHeroCoinOverlap(hero:Hero, thisCoin:Coin){
+		hero.updateScore(10);
+		score.text = hero.score+"";
+		thisCoin.kill();
+		coin = new Coin();
```

Comment:
Quick note on how `kill()` works - it simply stops update() and draw() calls from occurring for that object. It doesn't remove that object from memory or do any other kinds of cleanup. 

This means that the logic here creates a very, very small memory leak - theoretically a player could play long enough that so many coins have been created that the program exceeds its maximum permitted memory and crashes as a result.

Reply:
Rather than killing and creating a new one, you could just teleport the "consumed" coin to a new location, thus eliminating any need to allocate memory when the player touches a coin.